src/facets/MemeMarketFacet.sol

```diff
@@ -27,34 +31,59 @@ contract MemeMarketFacet is FacetBase {
         }
     }
 
-    function buyMeme(MemeBuySizeDollars size) external returns (uint160 sqrtPriceX96, uint256 priceX96, uint256 amount, uint256 buyAmount) {
+    function buyMeme(
+        MemeBuySizeDollars size
+    )
+        external
+        payable
+        nonReentrant
+        returns (uint160 sqrtPriceX96, uint256 priceX96, uint256 cost, uint256 buyAmount)
+    {
         AppStorage storage s = LibAppStorage.diamondStorage();
 
-        // get price of matic
+        // get price
         sqrtPriceX96 = LibUniswapV3Twap.getSqrtTwapX96(s.priceOracle, s.twapInterval);
 
         priceX96 = LibUniswapV3Twap.getPriceX96FromSqrtPriceX96(sqrtPriceX96);
 
         if (size == MemeBuySizeDollars.Five) {
             buyAmount = 1_000 ether;
-            amount = (priceX96 * 5e30) / FixedPoint96.Q96;
+            cost = (priceX96 * 5e30) / FixedPoint96.Q96;
+            // cost = uint256(5).divWadUp(1500);
         } else if (size == MemeBuySizeDollars.Ten) {
             buyAmount = 2_500 ether;
-            amount = (priceX96 * 10e30) / FixedPoint96.Q96;
+            cost = (priceX96 * 10e30) / FixedPoint96.Q96;
+            // cost = uint256(10).divWadUp(1500);
         } else if (size == MemeBuySizeDollars.Twenty) {
             buyAmount = 6_000 ether;
-            amount = (priceX96 * 20e30) / FixedPoint96.Q96;
+            cost = (priceX96 * 20e30) / FixedPoint96.Q96;
+            // cost = uint256(20).divWadUp(1500);
         } else if (size == MemeBuySizeDollars.Fifty) {
             buyAmount = 20_000 ether;
-            amount = (priceX96 * 50e30) / FixedPoint96.Q96;
+            cost = (priceX96 * 50e30) / FixedPoint96.Q96;
+            // cost = uint256(50).divWadUp(1500);
         } else if (size == MemeBuySizeDollars.Hundred) {
             buyAmount = 50_000 ether;
-            amount = (priceX96 * 100e30) / FixedPoint96.Q96;
+            cost = (priceX96 * 100e30) / FixedPoint96.Q96;
+            // cost = uint256(100).divWadUp(1500);
         }
 
+        require(msg.value >= cost, "Not enough coin sent");
+
         // first, transfer wmatic to treasury
-        IERC20 wmatic = IERC20(s.currencyAddress);
-        wmatic.transferFrom(_msgSender(), s.addresses[LibConstants.TREASURY_ADDRESS], amount);
+        // IERC20 wmatic = IERC20(s.currencyAddress);
+        // wmatic.transferFrom(_msgSender(), s.addresses[LibConstants.TREASURY_ADDRESS], amount);
+
+        // send coin to treasury, which is the server address
```

Comment:
Let's keep it as the treasury address constant for now, not to the server address directly. Our server address is a hot wallet and should only hold what's needed to transact, not our entire treasury. 

Instead, let's have a check function which, when called, checks if the server balance is below 1 eth and if so it then sends it some funds from the treasury. Let's call this new check function here at the bottom after receiving funds. Please make this check function public, so that anyone can call it at any time.

Reply:
Created LibComptroller._routeCoinPayment() to route coin payments to server address if the server has less than 0.5 eth, routes to the proxy otherwise.